Assets for AI/ML Profiling session (#54)
argonne-lcf/ALCF_Hands_on_HPC_Workshop: ai_ml_profiling/README.md

# Profiling a simple PyTorch example

> In the hands on session we will try to profile a simple PyTorch example. This mini-application (mini-benchmark) mimics the sequence parallelism strategy that we use for our auroraGPT training. A similar strategy is adopted by Megatron-DeepSpeed as well. 

Reviewer: ```suggestion
In the hands-on session we will try to profile a simple PyTorch example. This mini-application (or "mini-benchmark") mimics the sequence parallelism strategy that we use for our AuroraGPT training. A similar strategy is adopted by [Megatron-DeepSpeed](https://github.com/microsoft/Megatron-DeepSpeed) as well. 
```

> In the hands on session we will try to profile a simple PyTorch example. This mini-application (mini-benchmark) mimics the sequence parallelism strategy that we use for our auroraGPT training. Very similar strategy is adopted by Megatron-DeepSpeed as well. 

Author: Adding this.

> A request to an user would be to try and run the run the scripts either from an interactive session or a batch submission *before* jumping on to using the profilers! 

Reviewer: ```suggestion
Users should first attempt to run the following workload scripts without profiling on a Polaris compute node (from either an interactive session or a batch job submission).
```

> A request to an user would be to try and run the run the scripts both from an interactive session and a batch submission *before* jumping on to using the profilers! 

Author: It should be either! Fixing it!

    `ALLGATHER` --> `Matrix Multiplication 1` --> `Matrix Multiplication 2` --> `REDUCE_SCATTER`. In the docstring of the script, we see a hard coded "4". This comes from the assumption that, we have parallelized our matrix multiplication over 4 GPUs on a Polaris node.

- `qsub_polaris_sequence_parallelism.sh` \
    This is our job submission script. There are a few things that an user needs to modify.
    ```
    #PBS -l place=scatter
    #PBS -l walltime=00:05:00
    #PBS -q HandsOnHPC
    #PBS -l filesystems=home:grand
    #PBS -A alcf_training
    #PBS -N the_name_you_like
    #PBS -k doe
    #PBS -o /path/to/the/stdout
    #PBS -e /path/to/the/stderror

    ...
    WORK_DIR=/path/to/sequence_parallelism.py
    PROF_DIR=/path/to/the/directory/for/profiles
    ```

>     The `PROF_DIR` path is unused in this piece of code, as we have a separate script for launching the profiler.

Reviewer: why not delete this line from the script, then?

>     Keeping track of the stdout (`.OU`) and the stderror (`.ER`) files are important, as we plan to ask the NVIDIA Nsight Systems profiler to print out a summary for us. This summary comes in the .OU file.

Reviewer: ```suggestion
    Keeping track of the stdout (`.OU`) and the stderror (`.ER`) files are important, as we plan to ask the NVIDIA Nsight Systems profiler to print out a summary for us. This summary comes in the `.OU` file.
```

## Basic Strategy
The main code allows us to play around with a few parameters, like the number of elements in the GPU buffer (the sequence length, the number that goes in the sequence dimension of the input to a transformer layer stack), the hidden dimension (a proxy for the model dimension), and the precision type (only `float32` and `bfloat16` is supported). We will try to record results for however many combinations we can go through. The lower cutoff of expectation here is to at least get profiles for 1 and 2 nodes, for both precision type. We will also try to pay attention to the results coming out of the timing loops.

This will primarily be a two step process. Generate the profile on Polaris, `scp` or `rsync` it to the local machine (where we have `nsys` and `ncu` client/application/gui/viewer installed). Open the profiles and see if these make sense with respect to the workload that we have deployed.

Installing Nsight Systems and Nsight Compute is fairly straightforward. We can start by doing a "download Nsight Systems" google search. In any case, the two links at the bottom should take us to the desired location!

## Install Nsys

> 

Reviewer: Follow these links to install a version of NVIDIA's Nsight Systems and Nsight Compute on your local machine.

[Getting Started, Download Nsys](https://developer.nvidia.com/nsight-systems/get-started)

> [Download Nsight Compute](https://developer.nvidia.com/tools-overview/nsight-compute/get-started)

Reviewer: Would be nice to add a short section about `## View profiles` with some guiding hints on what to look for, what parameters you modify (unless you have that content planned for slides, etc.) Could also add a link to a public Box folder containing the `qdrep` etc. binary profile data for folks who are unable to collect the performance profiles themselves, for whatever reason.
